rfc/RFC-0001-dynamic-embedding-with-PS.md

```diff
@@ -0,0 +1,95 @@
+
+    Because TorchRec supports multiple sharding schema, different segments of an embedding feature may be stored on different devices, when given a `cache_id`, we need to cut it into segments and move those segments to their actual locations in the ShardedTensor.
+    We may store features in PS according to the sharding schema, i.e. save segments of the feature separately, so that each device can only fetch the segment that will be in its shard.
+
+2. How to have a consistent cache index across devices?
+
+    During training, the same `cache_id` on different devices must be storing the feature with the same `global_id`. To ensure strong consistency, we could use let each device only maintain the cache info of the features that are stored locally, and communicate with other devices if cache info of other features is needed.
+
+    <img src="./RFC-0001-assets/cache-consistency.png" style="width:400px;" />
+
+    In the illustration above, each GPU gets a different data sample (data-parallel training), and each data sample has some ids that should be stored on GPU 0, 1, and 2 (represented by blue, orange, and green). We could first use an all-to-all operation to gather all ids with the same color to its corresponding GPU so that each GPU could deal with their ids. And after updating the cache, we could use another all-to-all operation to scatter the ids back (not drawn in the above illustration).
+
+    **NOTE**: The actual cache calculation may be done on the CPU. GPU 0, 1, and 2 here merely represent the shardings.
+
+    The 2 unbalanced all-to-all may become a bottleneck of the system. We also proposed another design to reduce the communication:
+
+    <img src="./RFC-0001-assets/logical-cache-consistency.png" style="width:800px;" />
+
+    In the illustration above, all devices will first allgather the data. After obtaining all data in this step, we could simulate the changes in the cache of remote PS with a logical cache. Therefore, we could start 3 caches on each device, while only 1 of the 3 will update the embedding, the other ones are simulators. With these simulators, we no longer need to use another all-to-all operation to scatter the ids. And the allgather communication will be more balanced.
+
+    These 2 approaches are a trade-off between network communication and computation. It is worth mentioning that the latter approach is more complex. And the second all-to-all operation in the first approach could be concurrent with the fetching from PS to GPU cluster.
+
+3. Will there be intrusive modifications to TorchRec?
+
+    To implement prefetching, i.e. preprocessing future data while training on the current ones, we need a synchronization mechanism. This is because we need to know which data are ready to train and which have finished training (ready to evict). Also, we could postpone the synchronization of fetching to just before calling `forward` on embedding modules. We may register hooks to the DMP model.
```

Comment:
Basically like the graph below,

![image](https://user-images.githubusercontent.com/728699/176817506-e60a3607-4309-47db-9923-c0b087a239de.png)

There could be a background thread to preprocess data batch and fetch/evict correct parameters from/to parameter server. Since they are IO-intensive operators, they can be just asynchronous(e.g., python asyncio/C++ asio/CUDA memcpy async). In the background thread, we just launch asynchronous tasks and do not wait for them to complete until the data batch needs to be trained.

In another word, we just prefetch some data batches to pipeline `dynamic embedding` related I/O and normal training.